src/sources/pulsar.rs

```diff
@@ -0,0 +1,589 @@
+                consumer_builder.with_dead_letter_policy(pulsar::consumer::DeadLetterPolicy {
+                    max_redeliver_count: dead_letter_queue_policy.max_redeliver_count,
+                    dead_letter_topic: dead_letter_queue_policy.dead_letter_topic.clone(),
+                });
+        }
+
+        if let Some(batch_size) = self.batch_size {
+            consumer_builder = consumer_builder.with_batch_size(batch_size);
+        }
+        if let Some(consumer_name) = &self.consumer_name {
+            consumer_builder = consumer_builder.with_consumer_name(consumer_name);
+        }
+        if let Some(subscription_name) = &self.subscription_name {
+            consumer_builder = consumer_builder.with_subscription(subscription_name);
+        }
+
+        let consumer = consumer_builder.build::<String>().await?;
+
+        Ok(consumer)
+    }
+}
+
+async fn pulsar_source(
+    mut consumer: Consumer<String, TokioExecutor>,
+    decoder: Decoder,
+    mut shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    acknowledgements: bool,
+    log_namespace: LogNamespace,
+) -> Result<(), ()> {
+    let (finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, Some(shutdown.clone()));
+
+    let bytes_received = register!(BytesReceived::from(Protocol::TCP));
+
+    loop {
+        tokio::select! {
+            _ = &mut shutdown => break,
+            entry = ack_stream.next() => {
+                if let Some((status, entry)) = entry {
+                    handle_ack(&mut consumer, status, entry).await;
+                }
+            },
+            Some(maybe_message) = consumer.next() => {
+                match maybe_message {
+                    Ok(msg) => {
+                        bytes_received.emit(ByteSize(msg.payload.data.len()));
+                        parse_message(msg,&decoder,&finalizer,&mut out,&mut consumer,log_namespace).await;
+                    }
+                    Err(error) => {
+                        emit!(PulsarReadError { error })
+                    }
+                }
+            },
+        }
+    }
+
+    Ok(())
+}
+
+async fn parse_message(
+    msg: Message<String>,
+    decoder: &Decoder,
+    finalizer: &Option<OrderedFinalizer<FinalizerEntry>>,
+    out: &mut SourceSender,
+    consumer: &mut Consumer<String, TokioExecutor>,
+    log_namespace: LogNamespace,
+) {
+    let publish_time = i64::try_from(msg.payload.metadata.publish_time)
+        .ok()
+        .and_then(|millis| chrono::Utc.timestamp_millis_opt(millis).latest());
+    let topic = msg.topic.clone();
+    let producer_name = msg.payload.metadata.producer_name.clone();
+
+    let mut stream = FramedRead::new(msg.payload.data.as_ref(), decoder.clone());
+    let stream = async_stream::stream! {
+        while let Some(next) = stream.next().await {
+            match next {
+                Ok((events, _byte_size)) => {
+                    emit!(PulsarEventsReceived{
```

Comment:
Just to highlight one more metric - here we need to implement a new metric approach.